File: src/flourish/tracking/buff_tracker.py
from flourish.models.events import ApplyBuffEvent, RemoveBuffEvent, RefreshBuffEvent


class BuffTracker:
    def __init__(self):
        self._active: dict[int, int] = {}  # buff_id -> timestamp applied

    def process(self, event: ApplyBuffEvent | RemoveBuffEvent | RefreshBuffEvent):
        if isinstance(event, (ApplyBuffEvent, RefreshBuffEvent)):
            self._active[event.ability_id] = event.timestamp
        elif isinstance(event, RemoveBuffEvent):
            self._active.pop(event.ability_id, None)

    def is_active(self, buff_id: int) -> bool:
        return buff_id in self._active

    def get_applied_at(self, buff_id: int) -> int | None:
        return self._active.get(buff_id)

File: go-backend/internal/tracking/buff_tracker.go
package tracking

import "github.com/rdruid-talent-analyzer/go-backend/internal/models"

type BuffTracker struct {
	active map[int]int // buffID -> timestamp
}

func NewBuffTracker() *BuffTracker {
	return &BuffTracker{active: make(map[int]int)}
}

func (t *BuffTracker) Process(event models.Event) {
	switch e := event.(type) {
	case *models.ApplyBuffEvent:
		t.active[e.AbilityID] = e.Timestamp
	case *models.RefreshBuffEvent:
		t.active[e.AbilityID] = e.Timestamp
	case *models.RemoveBuffEvent:
		delete(t.active, e.AbilityID)
	}
}

func (t *BuffTracker) IsActive(buffID int) bool {
	_, ok := t.active[buffID]
	return ok
}

func (t *BuffTracker) GetAppliedAt(buffID int) *int {
	ts, ok := t.active[buffID]
	if !ok {
		return nil
	}
	return &ts
}
